src/algebra/module/polynomial.lean

```diff
@@ -0,0 +1,144 @@
+/-
+Copyright (c) 2022 Pierre-Alexandre Bazin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pierre-Alexandre Bazin
+-/
+import data.polynomial
+
+/-!
+# Modules over `R[X]`
+
+## Main definitions
+
+* `module_with_End E` : A type synonym for a `R`-module which explicitly mentions a given
+  endomorphism `E`. It naturally comes with a `R[X]`-module structure, where the action of `X` is
+  given by `E`.
+
+## Main statements
+
+* `module_with_End.polynomial_linear_equiv_semiconj` : equivalence between `R[X]`-linear maps and
+  `R`-linear maps compatible with the action of `X`. See also
+  `module_with_End.polynomial_lequiv_equiv_semiconj` for the restriction to linear isomorphisms.
+
+## Tags
+
+Module, polynomial
+-/
+
+
+variables {R M : Type*} [comm_semiring R] [add_comm_monoid M] [module R M]
+open_locale polynomial
+
+
+open polynomial module
+
+/--A type synonym for a `R`-module which explicitly mentions a given endomorphism.-/
+@[nolint unused_arguments has_inhabited_instance]
+def module_with_End (E : End R M) := M
+
+namespace module_with_End
+variables {E : End R M}
+
+instance : add_comm_monoid (module_with_End E) := (infer_instance : add_comm_monoid M)
+instance : module R (module_with_End E) := (infer_instance : module R M)
+/--The type conversion `M → module_with_End E`, as a `R`-linear isomorphism.-/
+def of_module (E : End R M) : M ≃ₗ[R] module_with_End E := linear_equiv.refl R M
+@[simp] lemma of_module_apply (u : M) : of_module E u = u := rfl
+instance apply_module : module (End R M) (module_with_End E) :=
+(infer_instance : module (End R M) M)
+
+/--Defines a `R[X]`-module structure on a `R`-module given an endomorphism `E` which defines the
+action of `X`.-/
+noncomputable instance polynomial_module : module R[X] (module_with_End E) :=
+comp_hom M (aeval E).to_ring_hom
+lemma smul_def (P : R[X]) (u : module_with_End E) : P • u = aeval E P u := rfl
+
+lemma X_smul (u : module_with_End E) : (X : R[X]) • u = E u :=
+by rw [smul_def, aeval_X]
```

Comment:
This lemma is mis-stated, `E` is a function from `M` not from `module_with_end E`:
```lean
lemma X_smul (u : module_with_End E) : (X : R[X]) • u = of_module E (E ((of_module E).symm u)) :=
by { rw [smul_def, aeval_X], refl }
```